### PR TITLE
Add automatic source routing support for DHCP/ipv6RA addrs

### PR DESCRIPTION
**Pull request by KJTsanaktsidis**
## TLDR
If you have multiple network interfaces on the same network, and want outgoing traffic to use a particular network interface based on the source address of the outgoing traffic, you need to made dedicated route tables for each interface and  configure routing policy rules. Networkd supports both putting DHCP/IPv6RA routes in specific tables (with `RouteTable=`), and supports making routing policy rules (with `[RoutingPolicyRule]`), but there's no way to automatically create a `[RoutingPolicyRule]` for each address an interface received from DHCP.

This PR adds two new sections `[IPv4SourceRouting]` and `[IPv6SourceRouting]` which will configure networkd to create routing policy rules for each address on an interface, including those received from DHCP.

I don't really have a good sense for what the config file interface for this should look like, so I'm happy to take any suggestions of better ways to make this configurable than the two new sections.

Thanks!

## Details
When there are multiple network interfaces on a machine attached to the same
network, Linux needs some help to work out when traffic should go out of each
interface, since addresses on Linux belong to the host, not the interface.

One way to configure this is with "source routing". In this scheme, you have
separate routing tables for each interface, each of which has an entry to send
traffic for the same local network out via that interface. Then, you use
routing policy rules to configure which routing table to use based on the
source address of a packet.

For example, if you had two interfaces eth1 and eth2 connected to the same
network, with addresses 10.0.0.2/24 and 10.0.0.3/24, and you wanted traffic
to use eth1 _unless_ its part of a TCP flow specifically directed at 10.0.0.3,
you might have route tables like so:

```
$ ip route list table 100
default via 10.0.0.1 dev eth1 proto dhcp src 10.0.0.2  metric 1024
10.0.0.0/24 dev eth1 proto dhcp scope link src 10.0.0.2
10.0.0.2 dev eth1 proto dhcp scope link src 10.0.0.2 metric 1024

$ ip route list table 101
default via 10.0.0.1 dev eth2 proto dhcp src 10.0.0.3  metric 1024
10.0.0.0/24 dev eth2 proto dhcp scope link src 10.0.0.3
10.0.0.3 dev eth2 proto dhcp scope link src 10.0.0.3 metric 1024

$ ip rule list
0:	from all lookup local
100:	from 10.0.0.2/24 lookup 100
101:	from 10.0.0./24 lookup 101
200:    from all lookup 100
32766:	from all lookup main
32767:	from all lookup default
```

Systemd already has support both for configuring DHCP routes in a
user-specified route table, and for configuring routing policy rules to achieve
this kind of source routing. However, there is no easy way to set this up in a
way that the addresses in the routing policy rules are automatically obtained
from DHCP/IPv6 router advertisements. That is what this patch tries to solve.

To use it in the scenario above, you would configure each interface like so:

```
[Match]
Name=eth1

[Network]
DHCP=yes
IPv6AcceptRA=yes

[DHCPv4]
RouteTable=100

[IPv6AcceptRA]
RouteTable=100

[RoutingPolicyRule]
Priority=200
Table=100
Family=ipv4

[RoutingPolicyRule]
Priority=200
Table=100
Family=ipv6

[IPv4SourceRoutingRules]
Enabled=yes
Priority=100

[IPv6SourceRoutingRules]
Enabled=yes
Priority=100
```

The `IPv{4,6}SourceRoutingRules` sections cause routing rules to be created for
each address on the network interface. For eth1 in our example above, we would
also want to create two RoutingPolicyRule entries to creat the "from all lookup
100" entry as well, so that outgoing traffic uses this interface. We would
_not_ want this for eth2.

### CODE REVIEW

**Comment by KJTsanaktsidis** [comment]
(this definitely still needs tests & code style cleanup I think, but I would love to get some feedback on the config interface before going back and adding them)

**Comment by topimiettinen** [comment]
@KJTsanaktsidis I can't comment on the network code very much in general, but perhaps this PR could be used to start implementing the Strong Host Model (#14397)? If so, it would be great if the configuration needed to enable the Strong Host Model would be simple.

**Comment by KJTsanaktsidis** [comment]
Right, strong host model is exactly what I'm trying to enable with this change. I read through #14397, your idea of having this be defined in `networkd.conf` is an interesting one - that this is really a conceptual setting about what host model you want, the routing policy rules and friends in here are just an implementation detail.

If it was controlled like that, we would probably want `HostModel=strong` to cause all DHCP obtained routes to be put in a routing table per interface, as if `RouteTable=` had been specified (or, perhaps, making it mandatory for this setting to be specified, since networkd wouldn't know what route table numbers are managed outside of networkd). Or, perhaps, we could have `StrongHostRouteTableRange=` specified in `networkd.conf` so administrators can specify what route table numbers can/can't be managed by networkd.

It's might an easier ask to bake this functionality in some kind of per-interface setting for a while, like the one I proposed in this PR, before promoting it to a global setting.

**Comment by KJTsanaktsidis** [comment]
Hi @yuwata 

Thanks for the review. So that we're clear,

* The settings to enable this feature, `EnableSourceRoutingRules=` and `SourceRoutingRulesPriority=`, should go on the `[DHCPv4]` and `[IPv6AcceptRA]` sections (happy to change the names if you have a better idea).
* I don't think this should be an option for `[DHCPv6]`, since it doesn't have a `RouteTable=` option to put the DHCP-obtained routes in another table. If we add that feature to `[DHCPv6]`, then we could add the source routing there too.
* It should only add source routing rules for addresses obtained via DHCP or IPv6RA, not any additional static addresses added with `Address=` lines.

Does that sound right to you? If so I'll start making these changes.

Thanks!

**Comment by yuwata** [comment]
> Does that sound right to you? If so I'll start making these changes.

Yes. Please.

**Comment by KJTsanaktsidis** [comment]
@yuwata I've made the changes (for IPv4 only), could you please take a look?

I haven't implemented it for IPv6 yet - if you're OK with the approach for IPv4 in here, I'll implement something similar in `networkd-ndisc.c` to handle the same config options in the `IPv6AcceptRA` section (which I think is the right place for this?)

Otherwise let me know what you think - I've never really used C in anger before so I'm sure there's plenty I've missed!

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        assert(link->dhcp4_messages > 0);
+
+        link->dhcp4_messages--;
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST) {
+                log_link_message_warning_errno(link, m, r, "Could not set DHCPv4 routing policy rule: %m");
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Please drop `: %m`. See #14834.

File: src/network/networkd-dhcp4.c [hunk 2]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+                const union in_addr_union address_union = { .in = address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+
+        /* Only actually add the rule if we didn't already add the same rule from a
+        previous lease. Otherwise, routing_policy_rule_configure will error with -EEXIST */
+        if (
+                link->dhcp4_source_routing_policy_rule &&
+                routing_policy_rule_compare_func(link->dhcp4_source_routing_policy_rule, rule)
+        ) {
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Coding style:
```
if (link->dhcp4_source_routing_policy_rule &&
    routing_policy_rule_compare_func(link->dhcp4_source_routing_policy_rule, rule)) {
```

File: src/network/networkd-dhcp4.c [hunk 3]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+                log_link_debug(link, "New and old source routing rules are the same, won't reconfigure");
+                return 0;
+        }
+        /* Now, if we really are going to set up a new routing policy rule, free the old one.
+        The dhcp4_remove_source_routing_policy_rule function will take care of actually calling
+        routing_policy_rule_remove based on the old lease address later on */
+        if (link->dhcp4_source_routing_policy_rule)
+                routing_policy_rule_free(TAKE_PTR(link->dhcp4_source_routing_policy_rule));
+
+
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Coding style: duplicated empty line.

File: src/network/networkd-dhcp4.c [hunk 4]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        log_link_debug(
+                link, "Adding source routing rule %s/%d -> * table %d pri %d family %d",
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Maybe, `%u` for prefixlen, and `pri` -> `priority`.

File: src/network/networkd-dhcp4.c [hunk 5]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        rule->from.in = *address;
+        rule->from_prefixlen = 32;
+        rule->family = AF_INET;
+        rule->priority = link->network->dhcp_source_routing_rule_priority;
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        if (!remove_all && link->dhcp4_source_routing_policy_rule &&
+                routing_policy_rule_compare_func(rule, link->dhcp4_source_routing_policy_rule) == 0
+        )
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Coding style: spurious indentation:
```
        if (!remove_all && link->dhcp4_source_routing_policy_rule &&
            routing_policy_rule_compare_func(rule, link->dhcp4_source_routing_policy_rule) == 0)
```

File: src/network/networkd-dhcp4.c [hunk 6]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+                return 0;
+
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = *address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+        log_link_debug(
+                link, "Removing source routing rule %s/%d -> * table %d pri %d family %d",
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Same here.

**Comment by yuwata** [comment]
Please also add a test case under `test/test-network`.

File: src/network/networkd-dhcp4.c (hunk 2 above)
Comment (KJTsanaktsidis) [reply src/network/networkd-dhcp4.c]: 👍  OK - I've indented like this in a few places, I'll fix them up.

File: src/network/networkd-dhcp4.c (hunk 4 above)
Comment (KJTsanaktsidis) [reply src/network/networkd-dhcp4.c]: Yup, I can tidy this up - I'll also replace `*` with `0.0.0.0/0` since I think that's what's used by the debug logging in `routing_policy_rule_configure.

File: src/network/networkd-dhcp4.c [hunk 7]
```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        log_link_debug(
+                link, "Adding source routing rule %s/%d -> * table %d pri %d family %d",
+                from_addr_str,
+                rule->from_prefixlen,
+                rule->table,
+                rule->priority,
+                rule->family
+        );
+
+        r = routing_policy_rule_configure(rule, link, dhcp4_routing_policy_rule_handler);
```
Comment (KJTsanaktsidis) [inline src/network/networkd-dhcp4.c]: Having thought about this for a while, I think rather than just blindly calling `routing_policy_rule_configure`, I should call `routing_policy_rule_get` first, and if it exists, call `routing_policy_rule_make_local`, like here: https://github.com/systemd/systemd/blob/5f04f4e470379f461b9f0f3e4f75e55cfc055536/src/network/networkd-link.c#L910

This is because:
    * `routing_policy_rule_configure` first sends the netlink message to configure the rule, _then_ adds it to its internal set of routing policy rules
    * If the rule already exists in the set, then `routing_policy_rule_configure` fails with `EEXIST`
    * But if that happens, `r <= 0` here is true, and we don't increment `link->dhcp4_messages++`, but
    * The netlink message was already sent, so it'll get a reply. This means that `link->dhcp4_messages` is out of sync with the actual number of outstanding netlink messages, and we're going to fail one of the `assert(link->dhcp4_messages > 0);` statements in one of the handlers

I think it's dangerous that `routing_policy_rule_configure` works this way - it would be better to wrap it in a function that provided its own netlink callback, and only invoked the provided callback if the original call returned success. However, that's probably a job for a different patch, so I'll do the `get` & `make_local` thing here for now? Sound good to you @yuwata ?

**Comment by KJTsanaktsidis** [comment]
Thanks for the review - I'll address the codestyle issues & then work on the IPv6 implementations for autoconf/dhcpv6 (I said earlier that DHCPv6 didn't have support for the `RouteTable=` directive, but I was wrong - it seems to inherit the value that's set for DHCPv4 - so I can implement this there as well).

**Comment by KJTsanaktsidis** [comment]
@yuwata @mrc0mmand I think this might be ready for another review - please take a look? 

* It now supports DHCPv6 and IPv6 SLAAC
* I've done my best to make sure the routing policy rules are cleaned up in the right circumstances
* I've added a couple of testcases
* Fixed the code style issues you identified previously

**Comment by KJTsanaktsidis** [comment]
@yuwata just wondering if you've had a moment to take another look? Thanks!

**Comment by KJTsanaktsidis** [comment]
Yeah ok I shouldn't have been lazy and tried to fix the merge conflicts in the Github UI :) I'll drop the merge commit & rebase when I get a moment.

File: man/systemd.network.xml [hunk 8]
```diff
@@ -1676,6 +1676,23 @@
...
           </listitem>
         </varlistentry>
+
+        <varlistentry>
+          <term><varname>EnableSourceRoutingRules=</varname></term>
+          <listitem>
+            <para>This option is used to configure source routing for an interface.
+            When combined with setting an explicit route table (with the <literal>RouteTable=</literal> option),
+            this causes routing policy rules to be automatically be created for the DHCP obtained address on the interface,
+            of the form "from $address lookup $routetable".</para>
```
Comment (poettering) [inline man/systemd.network.xml:1686]: hmm, this explains one low level language (our .network files), with another one (the iproute2 command line). It's only understandable if you know the letter well, and want to understand the former. I doubt this is necessarily a given though. Hence, can we express this in prose instead, please?

File: man/systemd.network.xml [hunk 9]
```diff
@@ -1676,6 +1676,23 @@
...
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <term><varname>SourceRoutingRulesPriority=</varname></term>
+          <listitem>
+            <para>When <literal>EnableSourceRoutingRules=</literal> is enabled, this controls the priority of the routing
+            policy rules that are created.</para>
```
Comment (poettering) [inline man/systemd.network.xml:1693]: please specify valid range, and whether smaller or larger values are more important. (alternatively point the reader to a relevant related option where this is explained)

File: src/network/networkd-dhcp4.c [hunk 10]
```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        rule->from.in = address;
+        rule->from_prefixlen = 32;
+        rule->family = AF_INET;
+        rule->priority = link->network->dhcp_source_routing_rule_priority;
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:638]: in_addr_to_string(AF_INET, (union in_addr_union) { .in = address }, &from_addr_str);

File: src/network/networkd-dhcp4.c [hunk 11]
```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        }
+
+        log_link_debug(
+                link, "Adding source routing rule %s/%u -> 0.0.0.0/0 table %d priority %d family %d",
+                from_addr_str,
+                rule->from_prefixlen,
+                rule->table,
+                rule->priority,
+                rule->family
+        );
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:648]: the `if` block checking for `DEBUG_LOGGING` should cover this `log_link_debug()` call too

File: src/network/networkd-dhcp4.c [hunk 12]
```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+
+        r = routing_policy_rule_get(link->manager, rule, &rrule);
+        if (r == 0) {
+            /* this means the rule already existed on the system, we just need to own it */
+            (void) routing_policy_rule_make_local(link->manager, rrule);
+        } else if (r < 0) {
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:654]: according to our coding style: no {} round single-line if blocks please

File: src/network/networkd-dhcp4.c [hunk 13]
```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        is_current_lease = link->dhcp4_source_routing_policy_rule &&
+            routing_policy_rule_compare_func(rule, link->dhcp4_source_routing_policy_rule) == 0;
+        if (!remove_all && is_current_lease)
+                return 0;
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = *address };
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:700]: as above

File: src/network/networkd-dhcp4.c [hunk 14]
```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+        log_link_debug(
+                link, "Removing source routing rule %s/%u -> 0.0.0.0/0 table %d priority %d family %d",
+                from_addr_str,
+                rule->from_prefixlen,
+                rule->table,
+                rule->priority,
+                rule->family
+        );
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:710]: as above.

(btw, using pretty unicode → instead of boring ascii -> is liked in the systemd project ;-))

File: src/network/networkd-dhcp4.c [hunk 15]
```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
...
+
+        (void) routing_policy_rule_remove(rule, link, NULL);
+        if (is_current_lease) {
+                routing_policy_rule_free(link->dhcp4_source_routing_policy_rule);
+                link->dhcp4_source_routing_policy_rule = NULL;
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:715]: hmm, we should rework routing_policy_rule_free() so that it returns `(RoutingPolicyRule*) NULL` unconditionally, so that we can write:

```c
link->dhcp4_source_routing_policy_rule = routing_policy_rule_free(link->dhcp4_source_routing_policy_rule);
```

we do things like that in almost all new code these days, and really should do this here too.

(of course, this is independent of this work of yours, but it would be very welcome to see a commit in your PR that converts this ;-))

File: man/systemd.network.xml (hunk 8 above)
Comment (poettering) [reply man/systemd.network.xml:1686]: should say "takes a boolean". also should say what the default is.

File: man/systemd.network.xml (hunk 9 above)
Comment (poettering) [reply man/systemd.network.xml:1693]: also, should say what the default is.

**Comment by poettering** [comment]
OK, let's close for now. if there's still ineterest, please rebase/refresh and post a new PR.

**Comment by KJTsanaktsidis** [comment]
Sorry - yes this fell off my list of things to do a long time ago. I’m sorry for having wasted your time with the review here, I really did mean to get around to working on this again… but yeah my focus has changed a lot since working on this.

**Comment by glance-** [comment]
I'd love to see this work (or something like https://github.com/systemd/systemd/pull/18275) finished and merged. How can I help to make that happen?
